File: src/tools/list-files.ts
import { readdir, stat } from "node:fs/promises";
import path from "node:path";

import type { ToolDefinition } from "../core/types.js";
import { safeDisplay } from "../text/safe-display.js";
import type { Workspace } from "../workspace/workspace.js";
import { isAgentVisibleName } from "../workspace/agent-path-policy.js";
import {
  objectSchema,
  optionalInteger,
  optionalString,
  parseArgumentObject,
} from "./arguments.js";
import { messageOf } from "./file-helpers.js";
import {
  type LocalTool,
  type PreparedToolCall,
  ToolInputError,
  failure,
  success,
} from "./types.js";

const IGNORED_DIRECTORIES = new Set(["node_modules"]);

export class ListFilesTool implements LocalTool {
  public readonly effect = "read" as const;
  public readonly definition: ToolDefinition = {
    name: "list_files",
    description:
      "List workspace files in deterministic order without following symlinked directories. Skips .git, .nash, and node_modules.",
    parameters: objectSchema({
      path: { type: "string", description: "Directory to list, default ." },
      max_depth: { type: "integer", minimum: 1, maximum: 5, description: "Recursive depth, default 2" },
      max_entries: { type: "integer", minimum: 1, maximum: 500, description: "Entry limit, default 200" },
    }),
  };

  readonly #workspace: Workspace;

  public constructor(workspace: Workspace) {
    this.#workspace = workspace;
  }

  public prepare(argumentsJson: string): PreparedToolCall {
    const object = parseArgumentObject(argumentsJson, ["path", "max_depth", "max_entries"]);
    const requestedPath = optionalString(object, "path", ".");
    if (requestedPath.trim() === "") {
      throw new ToolInputError("path must not be blank");
    }
    const maxDepth = optionalInteger(object, "max_depth", 2, 1, 5);
    const maxEntries = optionalInteger(object, "max_entries", 200, 1, 500);
    return {
      preview: `list ${safeDisplay(requestedPath)}`,
      execute: async (signal) =>
        await this.#execute(requestedPath, maxDepth, maxEntries, signal),
    };
  }

  async #execute(
    requestedPath: string,
    maxDepth: number,
    maxEntries: number,
    signal: AbortSignal,
  ) {
    if (signal.aborted) {
      return failure("listing was cancelled", "cancelled");
    }
    let root: string;
    try {
      root = await this.#workspace.resolveExisting(requestedPath);
    } catch (error) {
      return failure(`cannot resolve directory: ${messageOf(error)}`, "path_error");
    }

    try {
      if (!(await stat(root)).isDirectory()) {
        return failure("path is not a directory", "invalid_file_type");
      }
      const output: string[] = [];
      let entries = 0;
      let truncated = false;

      const visit = async (directory: string, depth: number): Promise<void> => {
        const children = await readdir(directory, { withFileTypes: true });
        children.sort((left, right) =>
          left.name < right.name ? -1 : left.name > right.name ? 1 : 0,
        );
        for (const child of children) {
          if (signal.aborted) {
            throw new CancellationError();
          }
          if (child.isDirectory() && IGNORED_DIRECTORIES.has(child.name)) {
            continue;
          }
          if (!isAgentVisibleName(child.name)) {
            continue;
          }
          if (entries >= maxEntries) {
            truncated = true;
            return;
          }
          const absolute = path.join(directory, child.name);
          let display = safeDisplay(this.#workspace.relative(absolute), 1_000);
          if (child.isDirectory()) {
            display += "/";
          } else if (child.isSymbolicLink()) {
            display += "@";
          }
          output.push(`${display}\n`);
          entries++;

          if (child.isDirectory() && depth < maxDepth) {
            await visit(absolute, depth + 1);
            if (truncated) {
              return;
            }
          }
        }
      };

      await visit(root, 1);
      if (entries === 0) {
        output.push("(directory is empty)\n");
      }
      if (truncated) {
        output.push(`... listing truncated after ${entries} entries ...\n`);
      }
      return success(output.join(""), { entries, truncated });
    } catch (error) {
      if (error instanceof CancellationError) {
        return failure("listing was cancelled", "cancelled");
      }
      return failure(`cannot list directory: ${messageOf(error)}`, "io_error");
    }
  }
}

class CancellationError extends Error {}

File: internal/tool/list_files.go
package tool

import (
	"context"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	"github.com/TmkShink/nash-agent/internal/core"
	"github.com/TmkShink/nash-agent/internal/workspace"
)

var errEntryLimit = errors.New("entry limit reached")

type ListFiles struct {
	workspace *workspace.Workspace
}

func NewListFiles(workspace *workspace.Workspace) *ListFiles {
	return &ListFiles{workspace: workspace}
}

func (t *ListFiles) Definition() core.ToolDefinition {
	return core.ToolDefinition{
		Name:        "list_files",
		Description: "List files and directories under a workspace-relative directory without following symlinked directories.",
		InputSchema: objectSchema(map[string]any{
			"path":        map[string]any{"type": "string", "description": "Directory to list, default ."},
			"max_depth":   map[string]any{"type": "integer", "minimum": 1, "maximum": 5, "description": "Maximum recursive depth, default 2"},
			"max_entries": map[string]any{"type": "integer", "minimum": 1, "maximum": 500, "description": "Maximum entries, default 200"},
		}),
	}
}

func (*ListFiles) Effect() Effect { return EffectRead }

func (*ListFiles) Preview(arguments string) string {
	var args struct {
		Path string `json:"path"`
	}
	if err := DecodeArguments(arguments, &args); err != nil || args.Path == "" {
		return "list ."
	}
	return "list " + args.Path
}

func (t *ListFiles) Execute(ctx context.Context, arguments string) Result {
	var args struct {
		Path       string `json:"path"`
		MaxDepth   int    `json:"max_depth"`
		MaxEntries int    `json:"max_entries"`
	}
	if err := DecodeArguments(arguments, &args); err != nil {
		return Failure("invalid list_files arguments: "+err.Error(), map[string]any{"kind": "invalid_arguments"})
	}
	if args.Path == "" {
		args.Path = "."
	}
	if args.MaxDepth == 0 {
		args.MaxDepth = 2
	}
	if args.MaxEntries == 0 {
		args.MaxEntries = 200
	}
	if args.MaxDepth < 1 || args.MaxDepth > 5 || args.MaxEntries < 1 || args.MaxEntries > 500 {
		return Failure("max_depth must be 1..5 and max_entries must be 1..500", map[string]any{"kind": "invalid_arguments"})
	}

	root, err := t.workspace.ResolveExisting(args.Path)
	if err != nil {
		return Failure("cannot resolve directory: "+err.Error(), map[string]any{"kind": "path_error"})
	}
	info, err := os.Stat(root)
	if err != nil {
		return Failure("cannot stat directory: "+err.Error(), map[string]any{"kind": "io_error"})
	}
	if !info.IsDir() {
		return Failure("path is not a directory", map[string]any{"kind": "invalid_file_type"})
	}
	var output strings.Builder
	entries := 0
	truncated := false
	err = filepath.WalkDir(root, func(path string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if err := ctx.Err(); err != nil {
			return err
		}
		if path == root {
			return nil
		}

		relFromRoot, err := filepath.Rel(root, path)
		if err != nil {
			return err
		}
		depth := strings.Count(filepath.ToSlash(relFromRoot), "/") + 1
		if depth > args.MaxDepth {
			if entry.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}
		if entry.IsDir() && (entry.Name() == ".git" || entry.Name() == ".nash") {
			return filepath.SkipDir
		}
		if entries >= args.MaxEntries {
			truncated = true
			return errEntryLimit
		}

		display := t.workspace.Relative(path)
		if entry.IsDir() {
			display += "/"
		} else if entry.Type()&fs.ModeSymlink != 0 {
			display += "@"
		}
		output.WriteString(display)
		output.WriteByte('\n')
		entries++
		return nil
	})
	if err != nil && !errors.Is(err, errEntryLimit) {
		return Failure("cannot list directory: "+err.Error(), map[string]any{"kind": "io_error"})
	}
	if entries == 0 {
		output.WriteString("(directory is empty)\n")
	}
	if truncated {
		output.WriteString(fmt.Sprintf("... listing truncated after %d entries ...\n", entries))
	}
	return Success(output.String(), map[string]any{"entries": entries, "truncated": truncated})
}
